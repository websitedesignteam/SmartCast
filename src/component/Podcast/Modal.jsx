import React from 'react';
import { Input } from "../../element"
import styles from "./Modal.module.scss";
import { baseUrl } from "../../utils/constants";

function Modal({
	input, 
	onChangeInput,
	isLoading, 
	onSubmitFavorite, 
	errorInputMessage,
	podcastDisclaimer,
	...props}) {

	return (
		
		<div className={styles.modalContainer}>
			<div className={styles.modalHeader}>
				Enter a nickname for your favorite podcast!
				<div className={styles.errorDisclaimer} >
					<img 
						className={styles.errorDisclaimerImage} 
						src={baseUrl + "/assets/button/disclaimer.svg"} 
						alt="disclaimer" 
					/>
					<div className={styles.errorDisclaimerText}>{podcastDisclaimer}</div>
				</div>
			</div>
			{(errorInputMessage) &&
				<div className={styles.error}>
					{errorInputMessage}
				</div>}
			<Input
				id="favoritePodcast-name" 
				name="name" 
				value={input} 
				label="" 
				placeholder="Enter Podcast Nickname" 
				onChangeInput={onChangeInput} 
			/>
			<div className={styles.footerContainer}>
				{!isLoading
					? <button className={styles.submit} onClick={onSubmitFavorite}>
				Confirm
			</button>
			: <div className="loaderTiny"></div>}
			</div>
		</div>
	);
}

export default Modal;