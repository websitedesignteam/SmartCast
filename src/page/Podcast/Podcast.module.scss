@import "../../styles/variables";

button {
  border: none;
}

.podcastContainer {
  padding-top: 2rem;
}

div.dataSection {
  margin-bottom: 1rem;
}

div.dataTitle {
  display: block;
}

div.error {
	font-size: 1.2rem;
}

.podcastImage {
  display: block;
  position: relative;
  margin: 0 auto 1rem;
  height: auto;
  width: 100%;
}

.episodeList {
  padding: 0;
  margin: 0;
  list-style-type: none;
  & li {
    margin-bottom: 0.5rem;
    width: 100%;
  }
  & a {
    display: block;
    text-decoration: none;
  }
  & li:hover {
    background-color: $light-gray;
  }
}

.episodePageButtons {
  display: flex;
  justify-content: center;
  button {
    border-radius: 100%;
    height: 2.5rem;
    width: 2.5rem;
    margin-right: 1rem;
    position: relative;
  }
  img {
    height: 1.5rem;
    width: 1.5rem;
    // margin: 0 0 -0.3rem -0.1rem;
    position: absolute;
    left: 50%;
    top: 50%; 
    transform: translate(-50%, -50%);
  }
}

.subHeader {
  display: flex;
  flex-direction: row;
  align-items: center;
}

.showMore {
  border-radius: 100%;
  margin-right: 0.4rem;
  padding: 0;
  height: 1rem;
  width: 1rem;

  > img.rotate {
    transform: rotate(45deg);
    -ms-transform: rotate(45deg);
    -webkit-transform: rotate(45deg);
    -moz-transform: rotate(45deg);
  }

  > img {
    height: 1rem;
    width: 1rem;
    margin: 0;
    padding: 0;
  }
}

.podcastTitle {
  display: inline-block;
  margin-right: 1rem;
  font-size: 1.25rem;
  margin-bottom: 0.5rem;
}

.podcastFavorite {
  background: transparent;
  padding: 0;
  margin-top: 0.2rem;
  position: relative;
  top: 0.3rem;
  height: 1.5rem;
  width: 1.5rem;
  img {
    height: 1.5rem;
    width: 1.5rem;
    margin: 0;
    padding: 0;
  }
}

button:hover:not([disabled]) {
  outline: none;
}

// button:focus {
//   outline: none;
//   border: none;
// }

.podcastFavorite:hover:not([disabled]) {
  background: none;
}

.podcastModal { 
  display: inline-block; 
  vertical-align: middle; 
  position: fixed;
  left: 50%;
  top: 50%; 
  transform: translate(-50%, -50%);
  z-index: 3;
}

.podcastContainer {
	@media screen and (min-width: 568px) {
		display: flex;
		align-items: stretch;
	}
}

.podcastHeader {
  display: flex;
  justify-content: space-between;
}

.desktopLeft {
	@media screen and (min-width: 568px) {
		border-right: 3px $gray solid;
		padding-right: 1.5rem;
	}

	@media screen and (min-width: 768px) {
    width: 30%;
    max-width: 30rem;
	}
}

.desktopRight {
	@media screen and (min-width: 568px) {
		padding-left: 1.5rem;
	}
}

.desktopRight, .desktopLeft {
	@media screen and (min-width: 568px) {
	}
}