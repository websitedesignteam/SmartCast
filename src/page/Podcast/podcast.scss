.podcast-container {
  margin: auto;
}

span {
  display: block;
}