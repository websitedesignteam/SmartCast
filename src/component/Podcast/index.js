import EpisodeCard from "./EpisodeCard/EpisodeCard"
import Modal from "./Modal/Modal"
import SectionContainer from "./SectionContainer/SectionContainer"

export {
    EpisodeCard,
    Modal,
    SectionContainer,
};

export default {
    EpisodeCard,
    Modal,
    SectionContainer,
};