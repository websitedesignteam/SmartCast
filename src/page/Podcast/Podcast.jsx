import React, { useState, useEffect, useRef } from 'react';
import { Link, useParams } from "react-router-dom";
import { getPodcast, postFavoritePodcast, getUser } from '../../utils/api';
import { useIsActive, useOnClickOutside } from 'hooks';
import styles from "./Podcast.module.scss";
import { isInFavoritePodcasts, getNameInFavoritePodcasts } from "../../utils/helper";
import { podcastCommands, podcastDisclaimer, errorFavoritePodcast, errorTooBusy } from "../../utils/constants"; 
import { Modal, EpisodeCard, SectionContainer } from "../../component/Podcast";
import { baseUrl, errorPodcast } from "../../utils/constants";

function Podcast({user, validateToken, setUser, ...props}) {
	//vars
	const { podcastID } = useParams();
	const { access_token, favoritePodcasts } = user;

	//states
	const [isLoading, setIsLoading] = useState(true);
	const [isLoadingFavorite, setIsLoadingFavorite] = useState(false);
	
	const [inputFavoritePodcast, setInputFavoritePodcast] = useState({podcastName: "", command: ""});
	const [currentPodcast, setCurrentPodcast] = useState(null);
	const [errorMessage, setErrorMessage] = useState(null);
	const showDescription = useIsActive(true);
	const showEpisodes = useIsActive(true);
	const favoritePodcast = useIsActive();
	const inputModalState = useIsActive();
	const [errorInputMessage, setErrorInputMessage] = useState("");

	//make into custom hook
	const [episodePageList, setEpisodePageList] = useState([]); 
	const [currentEpisodePageIndex, setCurrentEpisodePageIndex] = useState(-1);

	//refs
	const inputModalRef = useRef();
	useOnClickOutside(inputModalRef, () => {
		inputModalState.deactivate();
		setErrorInputMessage("");
	});

	//api call to get podcast metadata
	const getPodcastAPI = () => {
		const data = { podcastID, nextPage : episodePageList[currentEpisodePageIndex] }
		getPodcast(data)
		.then((response) => {
			setIsLoading(false);
			const podcastData = response.data.Data;
			setEpisodePageList(episodePageList.concat([podcastData.nextPageNumber]));
			const podcastCommand = isInFavoritePodcasts(podcastID, favoritePodcasts) ? "unfavorite" : "favorite";
			const podcastNameInFavorites = getNameInFavoritePodcasts(podcastID, favoritePodcasts);
			setInputFavoritePodcast({
				podcastName: podcastNameInFavorites ?? podcastData.podcastTitle,
				command: podcastCommands[podcastCommand],
			});
			(podcastCommand === "unfavorite") ? favoritePodcast.activate() : favoritePodcast.deactivate();
			setCurrentPodcast(podcastData);
		})
		.catch((error) => {
			console.log(error);
			if (error?.data?.message === "Too Many Requests") {
				alert(errorTooBusy);
			} else {
				setErrorMessage(errorPodcast);
			}
		});
	};

	const postFavoritePodcastAPI = () => {
		setIsLoadingFavorite(true);
		const data = {
			access_token,
			podcastID,
			...inputFavoritePodcast,
		}
		postFavoritePodcast(data)
		.then((response) => {
			setIsLoadingFavorite(false);
			const podcastCommand = (inputFavoritePodcast.command === "favorite") ? "unfavorite" : "favorite";
			setInputFavoritePodcast({
				...inputFavoritePodcast,
				command: podcastCommands[podcastCommand]
			})
			alert(response.data.Data);
			favoritePodcast.toggle();
			inputModalState.deactivate();
		})
		.catch((error) => {
			setIsLoadingFavorite(false);
			if (error?.data?.Error) {
				alert(error.data.Error);
			} else {
				alert(error);
			}
		})
	}

	const getUserAPI = () => {
		getUser({access_token})
        .then((response) => {
            const userData = response.data;
            const allUserData = { 
                ...user,
                ...userData, 
            }
            localStorage.setItem("user", JSON.stringify(allUserData));
            setUser(allUserData);
        })
        .catch((error) => {
        	console.log(error);
        });
	}

	//util functions
	const gotoLastPage = () => {
		if (currentEpisodePageIndex > -1) {
			setCurrentEpisodePageIndex(currentEpisodePageIndex-1)
		}
		else {
			return;
		}
	}

	const gotoNextPage = () => {
		if (currentEpisodePageIndex < episodePageList.length-1) {
			setCurrentEpisodePageIndex(currentEpisodePageIndex+1)
		}
		else {
			return;
		}
	}

	const onClickFavorite = () => {
		inputModalState.activate();
	}

	const onSubmitFavorite = () => {
		if (currentPodcast.podcastTitle !== inputFavoritePodcast.podcastName) {
			postFavoritePodcastAPI();
		}
		else {
			setErrorInputMessage(errorFavoritePodcast);
		}
	}

	function onChangeInput(event) {
        setInputFavoritePodcast({
			...inputFavoritePodcast, 
			podcastName: event.target.value
		});
    }

	useEffect(() => {
		if (!!access_token) {
			validateToken();
			getUserAPI();
		}
	}, [access_token]);

	useEffect(() => {
		getPodcastAPI();
	}, [currentEpisodePageIndex]);

	useEffect(() => {
        if (!isLoading) return;
		getPodcastAPI();
    }, [isLoading]);

	return (
		<>
			{(inputModalState.isActive) && 
			<div className={styles.podcastModal} ref={inputModalRef}>
				<Modal 
					input={inputFavoritePodcast.podcastName} 
					onChangeInput={onChangeInput} 
					isLoading={isLoadingFavorite} 
					onSubmitFavorite={onSubmitFavorite}
					errorInputMessage={errorInputMessage}
					podcastDisclaimer={podcastDisclaimer}
				/>
			</div>}
			<div className={styles.podcastContainer}>
				{ (currentPodcast && !errorMessage) 
				? <> 
					<div className={styles.desktopLeft}>
						<img className={styles.podcastImage} src={currentPodcast.podcastImage} alt="podcast cover" />
						<div className={styles.podcastHeader}>
							<div className={styles.podcastTitle}>
								<strong>{currentPodcast.podcastTitle}</strong>
							</div> 
							<button 
								className={styles.podcastFavorite} 
								onClick={favoritePodcast.isActive ? postFavoritePodcastAPI : onClickFavorite}
								title={(!access_token) 
									? "Sign in to add to your Favorites" 
									: favoritePodcast.isActive
									? "Remove from Favorites" 
									: "Favorite Podcast"
								}
								disabled={!access_token || !!isLoadingFavorite}
							>
								<img 
									src={baseUrl + ((favoritePodcast.isActive && inputFavoritePodcast["command"] === "remove") ? "/assets/button/heart-fill.svg" : "/assets/button/heart.svg")} 
									alt=""
								/>
							</button>
						</div>
					</div>

					<div className={styles.desktopRight}>
						<SectionContainer label="Publisher">
							{currentPodcast.podcastPublisher}
						</SectionContainer> 
						<div className={styles.podcastDescription}>
							<SectionContainer label="Description">
								{/* <div className={styles.subHeader}>
									<button className={styles.showMore} onClick={()=> showDescription.toggle()}>
										<img 
											src={baseUrl + "/assets/button/show-more.png"} 
											alt={showDescription.isActive ? "Hide Description" : "Show Description"}
											className={showDescription.isActive && styles.rotate} 
										/>
									</button>
									<strong>Description</strong> 
								</div> */}
								{ showDescription.isActive && 
								<p dangerouslySetInnerHTML={{__html: currentPodcast.podcastDescription}} /> }
							</SectionContainer>
						</div> 
						
						<div className={styles.podcastEpisodes}>
							<SectionContainer label="Episodes">
							{/* <div className={styles.subHeader}>
								<button className={styles.showMore} onClick={()=> showEpisodes.toggle()}>
									<img 
										src={baseUrl + "/assets/button/show-more.png"} 
										alt={showEpisodes.isActive ? "Hide Episodes" : "Show Episodes"}
										className={showEpisodes.isActive && styles.rotate}  
									/>
								</button>
								<strong>Episodes</strong> 
							</div> */}
							{ showEpisodes.isActive && 
							<>
								<ul className={styles.episodeList}>
								{ currentPodcast.episodes.map((episode, index) => 
									<li key={index} className={styles.episodeLink}>
										<Link to={`/podcast/${podcastID}/episode/${episode.episodeID}`}>
											<EpisodeCard episode={episode} />
										</Link>
									</li> 
								)}
								</ul> 

								<div className={styles.episodePageButtons}>
									{ currentEpisodePageIndex > -1 &&
									<button onClick={gotoLastPage}><img src={baseUrl + "/assets/button/page-back.png"} alt="Previous Page" title="Go to previous page"/></button>}
									{ currentEpisodePageIndex < episodePageList.length-1 &&
									<button onClick={gotoNextPage}><img src={baseUrl + "/assets/button/page-next.png"} alt="Next Page" title="Go to next page"/></button>}
								</div>
							</>}
							</SectionContainer>
						</div>
					</div>
				</>
				: (errorMessage) 
				? <div className={styles.error}>{errorMessage}</div>
				: <div className="loader" />
				}
			</div>
		</>
	);
}

export default Podcast;
