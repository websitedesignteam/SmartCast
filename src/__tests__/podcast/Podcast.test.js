import React from 'react';
import { shallow } from 'enzyme';
import Podcast from 'src/page/Podcast/Podcast';

function mockFunction() {
    const original = jest.requireActual('react-router');
    return {
      ...original,
      useParams: jest.fn().mockReturnValue({
        podcastID: "d5b2fe9bee744bc5b3b60fab397447b4"
      }),
    };
  }

jest.mock('react-router-dom', () => mockFunction());

test("should render the Podcast", () => {
    const validateToken = () => {}
    const setUser = () => {}
    const user = {}
    const props = {
        validateToken,
        setUser,
        user,
    }

    const match = { 
        params: { 
            podcastID: "d5b2fe9bee744bc5b3b60fab397447b4"
        } 
    }

    const component = shallow(
        <Podcast {...props} match={match} />
    )
    // expect(component.getElements()).toMatchSnapshot();
});