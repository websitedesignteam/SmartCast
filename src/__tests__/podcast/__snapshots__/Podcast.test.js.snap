// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the Podcast 1`] = `
Array [
  <React.Fragment>
    <div
      className="podcastContainer"
    >
      <div
        className="loader"
      />
    </div>
  </React.Fragment>,
]
`;
